internal/proxy/accesslog/log_writer.go

```diff
@@ -0,0 +1,352 @@
+	if err != nil {
+		return fmt.Errorf("can't open new logfile: %s", err)
+	}
+	l.file = f
+	l.size = 0
+	return nil
+}
+
+func (l *RotateLogger) closeFile() error {
+	if l.file == nil {
+		return nil
+	}
+	err := l.file.Close()
+	l.file = nil
+	return err
+}
+
+func (l *RotateLogger) retentionRunOnce() error {
+	files, err := l.oldLogFiles()
+	if err != nil {
+		return err
+	}
+
+	if l.maxBackups > 0 && l.maxBackups < len(files) {
+		for _, f := range files[l.maxBackups:] {
+			errRemove := os.Remove(path.Join(l.dir(), f.fileName))
+			if err == nil && errRemove != nil {
+				err = errRemove
+			}
+		}
+	}
+
+	return err
+}
+
+func (l *RotateLogger) retentionRun() {
+	defer l.closeWg.Done()
+	for {
+		select {
+		case <-l.closeCh:
+			log.Warn("close Access log retention")
+			return
+		case <-l.retentionCh:
+			_ = l.retentionRunOnce()
+		}
+	}
+}
+
+func (l *RotateLogger) retention() {
+	select {
+	case l.retentionCh <- true:
+	default:
+	}
+}
+
+func (l *RotateLogger) timeRotating() {
+	ticker := time.NewTicker(time.Duration(l.rotatedTime * int64(time.Second)))
+	log.Info("start time rotating of access log")
+	defer ticker.Stop()
+	defer l.closeWg.Done()
+
+	for {
+		select {
+		case <-l.closeCh:
+			log.Warn("close Access file logger")
+			return
+		case <-ticker.C:
+			l.Rotate()
+		}
+
+	}
+}
+
+//start rotate log file by time
+func (l *RotateLogger) start() {
+	l.closeCh = make(chan struct{})
+	l.closeWg = sync.WaitGroup{}
+	if l.rotatedTime > 0 {
+		l.closeWg.Add(1)
+		go l.timeRotating()
+	}
+
+	if l.maxBackups > 0 {
+		l.closeWg.Add(1)
+		l.retentionCh = make(chan bool, 1)
+		go l.retentionRun()
+	}
+}
+
+func (l *RotateLogger) max() int64 {
+	return int64(l.maxSize) * int64(megabyte)
+}
+
+func (l *RotateLogger) dir() string {
+	if l.localPath == "" {
+		return path.Join(os.TempDir(), "accesslog")
+	}
+	return l.localPath
+}
+
+func (l *RotateLogger) filename() string {
+	return path.Join(l.dir(), l.fileName)
+}
+
+func (l *RotateLogger) prefixAndExt() (string, string) {
+	ext := path.Ext(l.fileName)
+	prefix := l.fileName[:len(l.fileName)-len(ext)]
+	return prefix, ext
+}
+
+func (l *RotateLogger) newBackupName() string {
+	t := time.Now()
+	timestamp := t.Format(timeFormat)
+	name, suffix := l.prefixAndExt()
+	return path.Join(l.dir(), name+timestamp+suffix)
+}
+
+func (l *RotateLogger) oldLogFiles() ([]logInfo, error) {
+	files, err := ioutil.ReadDir(l.dir())
+	if err != nil {
+		return nil, fmt.Errorf("can't read log file directory: %s", err)
+	}
+
+	logFiles := []logInfo{}
+	prefix, ext := l.prefixAndExt()
+
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		if t, err := l.timeFromName(f.Name(), prefix, ext); err == nil {
+			logFiles = append(logFiles, logInfo{t, f.Name()})
+			continue
+		}
+	}
+
+	return logFiles, nil
+}
+
+func (l *RotateLogger) timeFromName(filename, prefix, ext string) (time.Time, error) {
+	if !strings.HasPrefix(filename, prefix) {
+		return time.Time{}, errors.New("mismatched prefix")
```

Comment:
add filename and prefix in error message

Reply:
This method just used for filtered the other file in the same directory with access log, and error info was not used.

Reply:
> add filename and prefix in error message

This method just used for filtered the other file in the same directory with access log, and error info was not used.